### PR TITLE
[EZP-29234] As a developer, I want to read about customizing REST API response

### PR DESCRIPTION
**Pull request by konradoboza**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [EZP-29234](https://jira.ez.no/browse/EZP-29234)
| Versions      | 1.7, 1.13, 2.1

This PR contains cookbook page which describes how to create custom REST API response. Example of such customization was introduced in https://github.com/ezsystems/PlatformUIBundle/pull/964.

### CODE REVIEW

File: docs/cookbook/creating_custom_rest_api_response.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    protected function visitVersionInfoAttributes(Visitor $visitor, Generator $generator, APIVersionInfo $versionInfo)
+    {
+        parent::visitVersionInfoAttributes($visitor, $generator, $versionInfo);
+
+        $this->visitTestValue($generator);
+    }
+
+    protected function visitTestValue(Generator $generator)
```
Comment (andrerom) [inline docs/cookbook/creating_custom_rest_api_response.md]: `visitCommentValue `

File: docs/cookbook/creating_custom_rest_api_response.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
+#Creating custom REST API response based on Accept header
+
+There are several situations when customizing REST API response could be useful. One of the examples is extending it by adding comments count to
```
[the remark this answers is not in the document]
Comment (konradoboza) [reply docs/cookbook/creating_custom_rest_api_response.md]: It's much better indeed, I was struggling a bit with the beginning, to be honest.

File: docs/cookbook/creating_custom_rest_api_response.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+class ValueObjectVisitorDispatcher extends BaseValueObjectVisitorDispatcher
+{
+    private $parentDispatcher;
+
+    public function __construct(BaseValueObjectVisitorDispatcher $parentDispatcher)
+    {
+        $this->parentDispatcher = $parentDispatcher;
+    }
+
+    public function setOutputVisitor(Visitor $outputVisitor)
```
Comment (bdunogier) [inline docs/cookbook/creating_custom_rest_api_response.md]: Not a big fan of these, but I guess that you couldn't find a simpler way to get this to work ?

File: docs/cookbook/creating_custom_rest_api_response.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+namespace Acme\ExampleBundle\Rest\Generator;
+
+use eZ\Publish\Core\REST\Common\Output\Generator\Json as BaseJson;
+
+class Json extends BaseJson
+{
+    protected function generateMediaType($name, $type)
+    {
+        return "application/vnd.ez.example.{$name}+{$type}";
```
Comment (bdunogier) [inline docs/cookbook/creating_custom_rest_api_response.md]: I don't understand that part. First, it feels odd that such an override is required. Then how does it work, as it uses the same code than the extended class: 

```
    public function getMediaType($name)
    {
        return $this->generateMediaType($name, 'json');
    }

    protected function generateMediaType($name, $type)
    {
        return "application/vnd.ez.api.{$name}+{$type}";
    }
```

Please help me :)

File: docs/cookbook/creating_custom_rest_api_response.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (konradoboza) [reply docs/cookbook/creating_custom_rest_api_response.md]: >Customizing REST API response can be useful in many situations. Both for headless and more traditional setups, REST responses can be enriched in a clean way and limit client to server round trips.

>To do this we can take advantage of eZ Platform's [HATEOAS](https://en.wikipedia.org/wiki/HATEOAS) based REST API and extend it with custom content types for our own needs. Specifically in this cookbook we'll add comments count to eZ\Publish\API\Repository\Values\Content\VersionInfo responses.

Are you ok with this version?

File: docs/cookbook/creating_custom_rest_api_response.md (hunk 4 above)
Comment (konradoboza) [reply docs/cookbook/creating_custom_rest_api_response.md]: It allows us to identify request by our custom `Accept` header, e.g. 'application/vnd.ez.example.VersionList+json'. 

The default one doesn't allow it by using *api* part:

```
protected function generateMediaType($name, $type)
{
    return "application/vnd.ez.api.{$name}+{$type}";
}
```

In this example I was able to replace it with *example*.

Comment (andrerom) [reply docs/cookbook/creating_custom_rest_api_response.md]: Maybe improve the naming a bit more, besides what Bertrand says maybe “my.api” as others should not use eZ as vendor name

File: docs/cookbook/creating_custom_rest_api_response.md (hunk 3 above)
Comment (andrerom) [reply docs/cookbook/creating_custom_rest_api_response.md]: Improvment can be done as a followup, given the doc here builds upon what has been released, it's imho good to go for now.

Maybe @adamwojs and/or @bdunogier can create a feature request to handle the improvment once there is priority for it.

**Comment by andrerom** [comment]
Status update: This can be merged, and @konradoboza plans to do a PR on master to simplify a bit on the steps which would then result in doc for master once that is being accepted.

**Comment by konradoboza** [comment]
@DominikaK do you want me to rebase it to get rid of the conflicts or do you prefer to do it differently?

**Comment by DominikaK** [comment]
@konradoboza yes, please :)

**Pull request merged**